### PR TITLE
[CUB] A04 Aleksei Beliaev

### PR DESCRIPTION
**Pull request by Alexeyha**

### CODE REVIEW

File: src/Stmt.lama [hunk 1]
```diff
@@ -19,8 +19,17 @@ import World;
...
-  failure ("Stmt eval not implemented\n")
+fun eval (c@[st, w], stmt) {
+  case stmt of 
+    Assn (v, expr) -> [ st <- [v, evalExpr(st, expr)], w ]
+  | Seq (stmt1, stmt2) -> eval(eval(c, stmt1), stmt2)
+  | Skip -> c
+  | Read (x) -> case readWorld(w) of rw@[value, new_w] -> [ st <- [x, value], new_w ] esac
+  | Write (expr) -> [st, writeWorld(evalExpr(st, expr), w)]
+  | If (expr, stmt1, stmt2) -> if (evalExpr(st, expr)) then eval(c, stmt1) else eval(c, stmt2) fi
+  | While (expr, stmt1) -> while (evalExpr(c.fst, expr)) do c := eval(c, stmt1) od; c
```
Comment (e2e4b6b7) [inline src/Stmt.lama:30]: nit: if + recursion is more funcional

**Review by e2e4b6b7 — CHANGES_REQUESTED** [review]

File: src/Stmt.lama (hunk 1 above)
Comment (Alexeyha) [reply src/Stmt.lama]: .... fixed

File: src/SM.lama [hunk 2]
```diff
@@ -137,6 +180,24 @@ public fun compileSM (stmt) {
...
+    | While (expr, s) -> 
+        case env.genLabels(2) of
+          [lab1, lab2, env1] -> case compile(lab, env1, s) of
+            [_, env, newCode] -> [false, env, singletonBuffer(JMP (lab1)) <+ LABEL (lab2) <+> newCode <+ LABEL (lab1) <+> compileExpr(expr) <+ CJMP ("nz", lab2) ]
+          esac
+        esac
+    | If (e, s1, s2) -> case env.genLabels(2) of 
+        [lab1, lab2, env] -> case compile(lab, env, s1) of
+          [_, env, s1Code] -> case compile(lab, env, s2) of
+            [_, env, s2Code] -> [false, env, compileExpr(e) <+ CJMP ("z", lab1) <+> s1Code <+ JMP (lab2) <+ LABEL (lab1) <+> s2Code <+ LABEL (lab2)]
```
[the remark this answers is not in the document]
Comment (Alexeyha) [reply src/SM.lama]: fixed

File: src/SM.lama [hunk 3]
```diff
@@ -137,6 +180,24 @@ public fun compileSM (stmt) {
                   esac
              esac
         esac
+    | While (expr, s) -> 
+        case env.genLabels(2) of
+          [lab1, lab2, env1] -> case compile(lab, env1, s) of
```
[the remark this answers is not in the document]
Comment (Alexeyha) [reply src/SM.lama]: fixed

File: src/SM.lama [hunk 4]
```diff
@@ -137,6 +180,24 @@ public fun compileSM (stmt) {
...
+        esac
+    | If (e, s1, s2) -> case env.genLabels(2) of 
+        [lab1, lab2, env] -> case compile(lab, env, s1) of
+          [_, env, s1Code] -> case compile(lab, env, s2) of
+            [_, env, s2Code] -> [false, env, compileExpr(e) <+ CJMP ("z", lab1) <+> s1Code <+ JMP (lab2) <+ LABEL (lab1) <+> s2Code <+ LABEL (lab2)]
+          esac 
+        esac 
+      esac 
+    | DoWhile (s, expr) -> case env.genLabel of 
+        [lab1, env] -> case compile(lab, env, s) of
```
[the remark this answers is not in the document]
Comment (Alexeyha) [reply src/SM.lama]: fixed

File: src/SM.lama [hunk 5]
```diff
@@ -182,20 +182,20 @@ public fun compileSM (stmt) {
...
-            [_, env, s2Code] -> [false, env, compileExpr(e) <+ CJMP ("z", lab1) <+> s1Code <+ JMP (lab2) <+ LABEL (lab1) <+> s2Code <+ LABEL (lab2)]
+            [_, env, s2Code] -> [true, env, compileExpr(e) <+ CJMP ("z", lab1) <+> s1Code <+ JMP (lab) <+ LABEL (lab1) <+> s2Code]
           esac 
         esac 
       esac 
     | DoWhile (s, expr) -> case env.genLabel of 
-        [lab1, env] -> case compile(lab, env, s) of 
-          [labUsed, env, sCode] -> [labUsed, env, singletonBuffer(LABEL (lab1)) <+> sCode <+> compileExpr(expr) <+ CJMP ("nz", lab1)]
+        [lab1, env] -> case compile(lab1, env, s) of 
+          [lab1Used, env, sCode] -> [lab1Used, env, singletonBuffer(LABEL (lab1)) <+> sCode <+> compileExpr(expr) <+ CJMP ("nz", lab1)]
```
Comment (e2e4b6b7) [inline src/SM.lama:198]: Now the end of body is the beginning of the body. You will skip the check of the condition for the end of loop and end up with the infinite loop

Also you are returning if the lab1 were used which is not the provided lab. Provided lab is unconditionally unused as for the while loop

**Review by e2e4b6b7 — CHANGES_REQUESTED** [review]

File: src/SM.lama (hunk 5 above)
Comment (Alexeyha) [reply src/SM.lama]: Oh yes, it was not a problem, when I didn't reuse labels in `if`. Thank you!!
Fixed

**Pull request closed without merging**